src/v/kafka/server/connection_context.cc

```diff
@@ -122,26 +122,54 @@ connection_context::dispatch_method_once(request_header hdr, size_t size) {
                     iobuf buf) mutable {
                 if (_rs.abort_requested()) {
                     // _proto._cntrl etc might not be alive
-                    return;
+                    return ss::now();
                 }
                 auto self = shared_from_this();
                 auto rctx = request_context(
                   self,
                   std::move(hdr),
                   std::move(buf),
                   sres.backpressure_delay);
-                // background process this one full request
+                /*
+                 * until authentication is complete, process requests in order
+                 * since all subsequent requests are dependent on authentication
+                 * having completed.
+                 *
+                 * the other important reason for disabling pipeling before
+                 * authentication completes is because when a sasl handshake
+                 * with version=0 is processed, the next data on the wire is
+                 * _not_ another request: it is a size-prefixed authentication
+                 * payload without a request envelope, and requires special
+                 * handling.
+                 *
+                 * a well behaved client should implicitly provide a data stream
+                 * that invokes this behavior in the server: that is, it won't
+                 * send auth data (or any other requests) until handshake or the
+                 * full auth-process completes, etc... but representing these
+                 * nuances of the protocol _explicitly_ in the server makes its
+                 * behavior easier to understand and avoids misbehaving clients
+                 * creating server-side errors that will appear as a corrupted
+                 * stream at best and at worst some odd behavior.
+                 */
+                auto f = do_process(std::move(rctx))
+                           .handle_exception([self](std::exception_ptr e) {
+                               vlog(
+                                 klog.info,
+                                 "Detected error processing request: {}",
+                                 e);
+                               self->_rs.conn->shutdown_input();
+                           })
+                           .finally([s = std::move(sres), self] {});
+
+                if (unlikely(!sasl().complete())) {
+                    return f;
+                }
+
                 (void)ss::with_gate(
                   _rs.conn_gate(),
-                  [this, rctx = std::move(rctx)]() mutable {
```

Comment:
wait, don't you need to handle the gate-closed exception too?

Reply:
I don't think so. I mean, maybe? This doesn't change any semantics wrt background processing of requests.

Reply:
It reads to me like the handle exception is in the gate future, no?

Reply:
ah, you're right it should be on the outside. but it won't handle gate-closed exception for this gate. that's what was confusing

Reply:
wouldn't `exception_ptr` capture all exceptions? including gate_closed_exception?

Reply:
with_gate throws gate closed, it doesn't return it as an exceptional future

Reply:
ack

Reply:
switched to try_gate_closed which does return exceptional future. so that also covers the case where on shutdown some request sneaks in. I've never seen that, but it should result now in a nicer log message if it happens.

Reply:
great! separately, i think that would coredump redpanda since it's not handled. i.e.: we need to log it and turn off the connection, but i think that's a separate task, not related to this PR.

Reply:
it is handled now by using try_gate_closed

Reply:
i guess you don't need the `self->_rs.conn->shutdown_input();`  because the dtor does it?

Reply:
my guess is that the original author of this thought that since all the background requests that are processed all have a shared pointer that keeps the connection context alive that the input should be shutdown _as soon as_ a problem is detected which might be quite a while before inflight operations finish the destructor runs. but I dunno if that is the exact life cycle concern here or not

Reply:
Sure. It’s also out of the scope of this PR